### PR TITLE
Fix single file code view detection on recent Phabricator versions

### LINKED ISSUE #4191: Diffusion code views are not detected on recent phabricator versions
1. Spin up a recent Phabricator instance using `./dev/phabricator/start.sh latest`
2. Add a repository
3. Navigate to a file on that repository
4. Witness no "View on Sourcegraph" button, no code intelligence, no "Code view added" message in the console.

### PR DESCRIPTION
**Pull request by lguychard**
Fixes #4191

### CODE REVIEW

File: browser/src/libs/phabricator/dom_functions.ts [hunk 1]
```diff
@@ -117,7 +117,19 @@ export const diffusionDOMFns: DOMFunctions = {
...
         }
-        return parseInt(lineAnchor.textContent || '', 10)
+        // In recent Phabricator versions, the line number is stored in the `data-n`
+        // attribute, and the textContent is empty.
+        if (lineAnchor.dataset.n !== undefined) {
+            const lineNumber = parseInt(lineAnchor.dataset.n, 10)
+            if (isNaN(lineNumber)) {
+                throw new Error('Could not parse lineNumber from data-n attribute')
+            }
+        }
```
Comment (chrismwendt) [inline browser/src/libs/phabricator/dom_functions.ts]: Did you forget a `return lineNumber`?

File: browser/src/libs/phabricator/dom_functions.ts [hunk 2]
```diff
@@ -117,7 +117,19 @@ export const diffusionDOMFns: DOMFunctions = {
         if (!lineAnchor) {
             throw new Error('could not find line number anchor from code element')
         }
-        return parseInt(lineAnchor.textContent || '', 10)
+        // In recent Phabricator versions, the line number is stored in the `data-n`
+        // attribute, and the textContent is empty.
+        if (lineAnchor.dataset.n !== undefined) {
+            const lineNumber = parseInt(lineAnchor.dataset.n, 10)
+            if (isNaN(lineNumber)) {
+                throw new Error('Could not parse lineNumber from data-n attribute')
```
Comment (chrismwendt) [inline browser/src/libs/phabricator/dom_functions.ts]:
```suggestion
                throw new Error(`Could not parse lineNumber from data-n attribute "${lineAnchor.dataset.n}"`)
```

File: browser/src/libs/phabricator/dom_functions.ts [hunk 3]
```diff
@@ -117,7 +117,19 @@ export const diffusionDOMFns: DOMFunctions = {
...
+            }
+        }
+        const lineNumber = parseInt(lineAnchor.textContent || '', 10)
+        if (isNaN(lineNumber)) {
+            throw new Error('Could not parse lineNumber from lineAnchor textContent')
```
Comment (chrismwendt) [inline browser/src/libs/phabricator/dom_functions.ts]:
```suggestion
            throw new Error(`Could not parse lineNumber from lineAnchor textContent "${lineAnchor.textContent}"`)
```

File: browser/src/libs/phabricator/dom_functions.ts (hunk 1 above)
Comment (lguychard) [reply browser/src/libs/phabricator/dom_functions.ts]: I sure did, thanks for catching!

File: browser/src/libs/phabricator/code_intelligence.ts [hunk 4]
```diff
@@ -157,10 +157,20 @@ const diffusionSourceCodeViewResolver = toCodeViewResolver('.diffusion-source',
     toolbarButtonProps,
 })
 
+// Matches Diffusion single file code views on recent Phabricator versions.
```
Comment (felixfbecker) [inline browser/src/libs/phabricator/code_intelligence.ts]: Would be good to also add a comment to the other resolvers that they are *only* for old versions. Having resolvers for different versions can make things really hard to understand when writing tests, fixing bugs or updating fixtures

**Pull request merged**